### PR TITLE
Switch from fastai to pytorch-lightning

### PR DESCRIPTION
**Pull request by leifdenby**
- [x] `pytorch.Dataset` for handling loading of individual triplets
- [x] `pytorch_lightning.LightningDataModule` for handling transform, batching and splitting data for train/test
   - [x] enable constructing backbone from those available in `torchvision` (done using `lightning-flash`)
     - <del> enable use of pre-trained weights even when input doesn't have the same number of channels as the loaded backbone. fastai does this by averaging/assigning the weights to the first few channels: https://github.com/fastai/fastai/blob/master/fastai/vision/learner.py#L33, could be nice to have something like this, but it should probably be generalised </del> - I'll do this as a separate feature. I think this functionality might be nice factored out in its own module
   - [x] add same transforms as we had previously with fastai
      - this looks a bit challenging, there are a lot to consider. https://github.com/muellerzr/fastinference_pytorch/tree/master appears to re-implement fastai's transforms in pytorch Transform classes
   - [x] ensure we're using the right normalisation when using pretrained network
- [x] `pytorch_lightning.LightningModule` which handles the training logic: building model, setting learning rate, batch size
- [x] refactor code for downloading and unpacking example datasets
- [x] refactor `get_embeddings` function to work with new architecture
- [x] run training on <del>full 10k</del> 5k examples dataset and inspect results
- [x] refactor inference plots to ensure they work with the new architecture
- [x] implement loading of models
- [x] set up automated tests on windows
- [x] set up automated tests on linux
- [ ] reactor triplet-dataset creation pipeline
- [ ] refactor `rectpred` EUREC4A code for making predictions on a rectangular domain
    - [ ] domain data creation
    - [ ] inference (making embeddings by applying NN on sliding window)
    - [ ] RGB composite plotting
- [x] Add ARC3 HPC system submission script (SGE) for running training on single GPU using local storage
- [x] implement dataloader which loads all images in dataset into RAM on init to try and avoid disc-access bottleneck

A nice outcome of implementing the `LightningModule` is that I've written code to accept input with *any* number of input channels now @MarkUoLeeds :smile: That will make it quite easy to work on say gridded data stored in netCDF files with more than just 3 "channels" (I just need to write another `pytorch.Dataset` for that and work out how to normalise the inputs.

https://github.com/leifdenby/convml_tt/issues/16

### CODE REVIEW

**Comment by leifdenby** [comment]
For my own reference this is what the network used to look like when created with fastai:
```
Sequential(
  (0): Sequential(
    (0): Conv2d(3, 64, kernel_size=(7, 7), stride=(2, 2), padding=(3, 3), bias=False)
    (1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (2): ReLU(inplace)
    (3): MaxPool2d(kernel_size=3, stride=2, padding=1, dilation=1, ceil_mode=False)
    (4): Sequential(
      (0): BasicBlock(
        (conv1): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
        (bn1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (relu): ReLU(inplace)
        (conv2): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
        (bn2): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      )
      (1): BasicBlock(
        (conv1): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
        (bn1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (relu): ReLU(inplace)
        (conv2): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
        (bn2): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      )
    )
    (5): Sequential(
      (0): BasicBlock(
        (conv1): Conv2d(64, 128, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False)
        (bn1): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (relu): ReLU(inplace)
        (conv2): Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
        (bn2): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (downsample): Sequential(
          (0): Conv2d(64, 128, kernel_size=(1, 1), stride=(2, 2), bias=False)
          (1): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        )
      )
      (1): BasicBlock(
        (conv1): Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
        (bn1): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (relu): ReLU(inplace)
        (conv2): Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
        (bn2): BatchNorm2d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      )
    )
    (6): Sequential(
      (0): BasicBlock(
        (conv1): Conv2d(128, 256, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False)
        (bn1): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (relu): ReLU(inplace)
        (conv2): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
        (bn2): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (downsample): Sequential(
          (0): Conv2d(128, 256, kernel_size=(1, 1), stride=(2, 2), bias=False)
          (1): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        )
      )
      (1): BasicBlock(
        (conv1): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
        (bn1): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (relu): ReLU(inplace)
        (conv2): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
        (bn2): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      )
    )
    (7): Sequential(
      (0): BasicBlock(
        (conv1): Conv2d(256, 512, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False)
        (bn1): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (relu): ReLU(inplace)
        (conv2): Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
        (bn2): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (downsample): Sequential(
          (0): Conv2d(256, 512, kernel_size=(1, 1), stride=(2, 2), bias=False)
          (1): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        )
      )
      (1): BasicBlock(
        (conv1): Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
        (bn1): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (relu): ReLU(inplace)
        (conv2): Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
        (bn2): BatchNorm2d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      )
    )
  )
  (1): Sequential(
    (0): AdaptiveConcatPool2d(
      (ap): AdaptiveAvgPool2d(output_size=1)
      (mp): AdaptiveMaxPool2d(output_size=1)
    )
    (1): Flatten()
    (2): BatchNorm1d(1024, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (3): Dropout(p=0.25)
    (4): Linear(in_features=1024, out_features=512, bias=True)
    (5): ReLU(inplace)
    (6): BatchNorm1d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (7): Dropout(p=0.5)
    (8): Linear(in_features=512, out_features=100, bias=True)
  )
)
```

It's made from:
1) a resnet18 encoder to extract image featuers
2) a 2D "adaptive concat-pool" layer (which combines adaptive max and avg pool) to produce a tensor with `(batch_size, n_resnet_features, 1, 1)`

**Comment by leifdenby** [comment]
Transforms originally applied in within fastai:

```python
def get_transforms(do_flip:bool=True, flip_vert:bool=False, max_rotate:float=10., max_zoom:float=1.1,
                   max_lighting:float=0.2, max_warp:float=0.2, p_affine:float=0.75,
                   p_lighting:float=0.75, xtra_tfms:Optional[Collection[Transform]]=None)->Collection[Transform]:
```

I was calling this as:

```python
        .transform(fastai.vision.get_transforms(flip_vert=True,))
```

**Comment by leifdenby** [comment]
Here's a run of training (fine-tuning) a resnet18 on a dataset with 2k triplets with `margin=10`:

convml-train-N2000

The loss doesn't go all the way to zero, but it didn't when I originally trained the model and there is a marked difference in the "near" and "distant" scales which is great. I think this is good enough for now :smile:

**Comment by MarkUoLeeds** [comment]
phew!
